extensions/amp-image-viewer/0.1/amp-image-viewer.js

```diff
@@ -261,104 +245,120 @@ export class AmpImageViewer extends AMP.BaseElement {
   }
 
   /**
-   * @return {number}
-   * @private
-   */
-  getSourceWidth_() {
-    if (this.sourceAmpImage_.hasAttribute('width')) {
-      return parseInt(this.sourceAmpImage_.getAttribute('width'), 10);
-    } else {
-      const img = elementByTag(dev().assertElement(this.sourceAmpImage_),
-          'img');
-      return img ? img.naturalWidth : this.sourceAmpImage_./*OK*/offsetWidth;
-    }
-  }
-
-  /**
-   * @return {number}
+   * Sets the source width and height based on the user-defined dimensions of
+   * the amp-img if exists, the natural dimensions of the source image if
+   * loaded, and the offset dimensions of amp-img element if not.
+   * @param {!Element} ampImg
    * @private
    */
-  getSourceHeight_() {
-    if (this.sourceAmpImage_.hasAttribute('height')) {
-      return parseInt(this.sourceAmpImage_.getAttribute('height'), 10);
+  setSourceDimensions_(ampImg) {
+    const width = ampImg.getAttribute('width');
+    const height = ampImg.getAttribute('height');
+    if (width && height) {
+      this.sourceWidth_ = width;
+      this.sourceHeight_ = height;
     } else {
-      const img = elementByTag(dev().assertElement(this.sourceAmpImage_),
-          'img');
-      return img ? img.naturalHeight : this.sourceAmpImage_./*OK*/offsetHeight;
+      const img = elementByTag(ampImg, 'img');
+      if (img) {
+        this.sourceWidth_ = img.naturalWidth || ampImg./*OK*/offsetWidth;
+        this.sourceHeight_ = img.naturalHeight || ampImg./*OK*/offsetHeight;
+      } else {
+        this.sourceWidth_ = ampImg./*OK*/offsetWidth;
+        this.sourceHeight_ = ampImg./*OK*/offsetHeight;
+      }
     }
   }
 
   /**
    * Initializes the image viewer to the target image element such as
    * "amp-img". The target image element may or may not yet have the img
    * element initialized.
+   * @returns {!Promise}
    */
   init_() {
-    this.sourceWidth_ = this.getSourceWidth_();
-    this.sourceHeight_ = this.getSourceHeight_();
-    this.srcset_ = srcsetFromElement(dev().assertElement(this.sourceAmpImage_));
+    if (this.image_) {
+      return Promise.resolve();
+    }
 
-    ARIA_ATTRIBUTES.forEach(key => {
-      if (this.sourceAmpImage_.hasAttribute(key)) {
-        this.image_.setAttribute(key, this.sourceAmpImage_.getAttribute(key));
-      }
-    });
-    st.setStyles(dev().assertElement(this.image_), {
-      top: st.px(0),
-      left: st.px(0),
-      width: st.px(0),
-      height: st.px(0),
+    this.image_ = this.element.ownerDocument.createElement('img');
+    this.image_.classList.add('i-amphtml-image-viewer-image');
+    const ampImg = dev().assertElement(this.sourceAmpImage_);
+    this.setSourceDimensions_(ampImg);
+    this.srcset_ = srcsetFromElement(ampImg);
```

Comment:
Is it possible to use the srcset already created by the img?

Reply:
The `img` that is a child of `amp-img` is not guaranteed to have a srcset (it might just have a src). The main purpose of `srcsetFromElement` here is to convert `src | srcset` into a `srcset` so we can use it later in the `srcset` logic. I don't think this is very good code, but the entire `srcset` logic in this component needs to undergo refactoring after we properly implement `srcset` and `sizes` anyway, so my preference is just to leave it as is. This PR doesn't touch the `srcset` logic, other than moving pieces of code around.